instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java

```diff
@@ -54,189 +43,82 @@
-
-    endpointPeerServiceMapping = getMapSettingFromEnvironment(ENDPOINT_PEER_SERVICE_MAPPING);
-
-    log.debug("New instance: {}", this);
+  /**
+   * Sets the agent configuration singleton. This method is only supposed to be called once, from
+   * the agent classloader just before the first instrumentation is loaded (and before {@link
+   * Config#get()} is used for the first time).
+   */
+  public static void internalInitializeConfig(Config config) {
+    if (INSTANCE != DEFAULT) {
+      log.warn("Config#INSTANCE was already set earlier");
+      return;
+    }
+    INSTANCE = requireNonNull(config);
   }
 
-  // Read order: Properties -> Parent
-  private Config(Properties properties, Config parent) {
-    exporterJar = properties.getProperty(EXPORTER_JAR, parent.exporterJar);
-    exporter = properties.getProperty(EXPORTER, parent.exporter);
-
-    propagators = getPropertyListValue(properties, PROPAGATORS, parent.propagators);
-
-    traceEnabled = getPropertyBooleanValue(properties, TRACE_ENABLED, parent.traceEnabled);
-    integrationsEnabled =
-        getPropertyBooleanValue(properties, INTEGRATIONS_ENABLED, parent.integrationsEnabled);
-
-    excludedClasses =
-        getPropertyListValue(properties, TRACE_CLASSES_EXCLUDE, parent.excludedClasses);
-
-    scopeDepthLimit =
-        getPropertyIntegerValue(properties, SCOPE_DEPTH_LIMIT, parent.scopeDepthLimit);
-
-    runtimeContextFieldInjection =
-        getPropertyBooleanValue(
-            properties, RUNTIME_CONTEXT_FIELD_INJECTION, parent.runtimeContextFieldInjection);
-
-    traceAnnotations = properties.getProperty(TRACE_ANNOTATIONS, parent.traceAnnotations);
-
-    traceMethods = properties.getProperty(TRACE_METHODS, parent.traceMethods);
-    traceAnnotatedMethodsExclude =
-        properties.getProperty(
-            TRACE_ANNOTATED_METHODS_EXCLUDE, parent.traceAnnotatedMethodsExclude);
-
-    traceExecutorsAll =
-        getPropertyBooleanValue(properties, TRACE_EXECUTORS_ALL, parent.traceExecutorsAll);
-    traceExecutors = getPropertyListValue(properties, TRACE_EXECUTORS, parent.traceExecutors);
-
-    sqlNormalizerEnabled =
-        getPropertyBooleanValue(properties, SQL_NORMALIZER_ENABLED, parent.sqlNormalizerEnabled);
-
-    kafkaClientPropagationEnabled =
-        getPropertyBooleanValue(
-            properties, KAFKA_CLIENT_PROPAGATION_ENABLED, parent.kafkaClientPropagationEnabled);
+  public static Config get() {
+    return INSTANCE;
+  }
 
-    hystrixTagsEnabled =
-        getBooleanSettingFromEnvironment(HYSTRIX_TAGS_ENABLED, parent.hystrixTagsEnabled);
+  public abstract Map<String, String> getAllProperties();
 
-    endpointPeerServiceMapping =
-        getPropertyMapValue(
-            properties, ENDPOINT_PEER_SERVICE_MAPPING, parent.endpointPeerServiceMapping);
+  @Nullable
+  public String getProperty(String propertyName) {
+    return getAllProperties().get(normalizePropertyName(propertyName));
```

Comment:
i think it's only integration names that need to be normalized
```suggestion
    return getAllProperties().get(propertyName);
```

Reply:
Yes, only integration names -- but we have to normalize them both when building the config (since someone can use dash in e.g. system property `-Dotel.integration.some-library.enabled=false`) and when `isIntegrationEnabled()` is called.
So I opted for adding the normalisation everywhere to be 100% consistent.

Reply:
makes sense 👍